driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -86,37 +90,54 @@ public void run() {
             }
         };
 
+        this.connections = new CopyOnWriteArrayList<PooledConnection>();
+        this.open = new AtomicInteger();
+    }
+
+    public ListenableFuture<Void> initAsync() {
         // Create initial core connections
-        List<PooledConnection> l = new ArrayList<PooledConnection>(options().getCoreConnectionsPerHost(hostDistance));
-        try {
-            for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++)
-                l.add(manager.connectionFactory().open(this));
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            // If asked to interrupt, we can skip opening core connections, the pool will still work.
-            // But we ignore otherwise cause I'm not sure we can do much better currently.
-        } catch (ConnectionException e) {
-            forceClose(l);
-            throw e;
-        } catch (UnsupportedProtocolVersionException e) {
-            forceClose(l);
-            throw e;
-        } catch (ClusterNameMismatchException e) {
-            forceClose(l);
-            throw e;
-        } catch (RuntimeException e) {
-            forceClose(l);
-            throw e;
+        int capacity = options().getCoreConnectionsPerHost(hostDistance);
+        final List<PooledConnection> connections = Lists.newArrayListWithCapacity(capacity);
+        final List<ListenableFuture<Void>> connectionFutures = Lists.newArrayListWithCapacity(capacity);
+        for (int i = 0; i < options().getCoreConnectionsPerHost(hostDistance); i++) {
+            PooledConnection connection = manager.connectionFactory().newConnection(this);
+            connections.add(connection);
+            connectionFutures.add(connection.initAsync());
         }
-        this.connections = new CopyOnWriteArrayList<PooledConnection>(l);
-        this.open = new AtomicInteger(connections.size());
 
-        logger.trace("Created connection pool to host {}", host);
+        ListenableFuture<List<Void>> allConnectionsFuture = Futures.allAsList(connectionFutures);
+
+        final SettableFuture<Void> initFuture = SettableFuture.create();
+        Futures.addCallback(allConnectionsFuture, new FutureCallback<List<Void>>() {
+            @Override
+            public void onSuccess(List<Void> l) {
+                // If the pool was closed, close connections and raise an exception when Close completes.
+                if (isClosed()) {
+                    phase = Phase.INIT_FAILED;
+                    initFuture.setException(new ConnectionException(host.getSocketAddress(), "Pool was closed."));
+                    forceClose(connections);
+                } else {
+                    HostConnectionPool.this.connections.addAll(connections);
+                    open.set(l.size());
+                    logger.trace("Created connection pool to host {}", host);
+                    phase = Phase.READY;
+                    initFuture.set(null);
```

Comment:
There might still be a race if the pool was closed after the check. Will take a closer look at it.